File: CrackingTheCodingInterview/ch04-TreesAndGraphs/buildOrder.js
/*
  given list of projects and dependencies, find a valid build order

  input:
    projects: a, b, c, d, e, f
    dependencies: (a, d), (f, b), (b, d), (f, a), (d, c)
  output: f, e, a, b, d, c
*/

function buildOrder(projects, dependencies) {
  const adjacencyList = buildAdjacencyList(projects, dependencies);
  const buildOrder = [];


  const DFS = (node) => {
    const edges = adjacencyList[node].edges;
    if (edges.length === 0 || edges.every(e => adjacencyList[e].built)) {
      if (!adjacencyList[node].built) {
        buildOrder.push(node);
        adjacencyList[node].built = true;
      }
    } else {
      edges.forEach(e => {
        if (!adjacencyList[e].built) DFS(e);
      })
    }
  };

  let i = 0;
  while (buildOrder.length < projects.length) {
    DFS(projects[i]);

    if (i === projects.length - 1) {
      i = 0;
    } else {
      i++;
    }
  }

  return buildOrder.reverse();
}

function buildAdjacencyList(projects, dependencies) {
  const list = {};
  projects.forEach(proj => {
    list[proj] = {
      built: false,
      edges: []
    };
  });

  dependencies.forEach(dep => {
    list[dep[0]].edges.push(dep[1]);
  });

  return list;
}

const projects = ['a', 'b', 'c', 'd', 'e', 'f'];
const dependencies = [['a', 'd'], ['f', 'b'], ['b', 'd'], ['f', 'a'], ['d', 'c']];

console.log(buildOrder(projects, dependencies));

File: CrackingTheCodingInterview/ch04-TreesAndGraphs/buildOrder.py
"""
Given list of projects and dependencies, find a valid build order

Input:
Projects: [a, b, c, d, e, f]
Dependencies: [(a, d), (f, b), (b, d), (f, a), (d, c)] (a is a dependency of f)
Output: f, e, a, b, d, c
"""


def build_order(projects, dependencies):
    build = []
    a = build_adjacency_list(projects, dependencies)

    def search(node):
        if a[node]['built']:
            return

        # If node does not have dep or the dep have already been built
        if len(a[node]['edges']) == 0 or all([a[edge]['built'] for edge in a[node]['edges']]):
            build.append(node)
            a[node]['built'] = True
        # Else DFS to find an edge that can be built
        else:
            for edge in a[node]['edges']:
                search(edge)

    # Driver
    i = 0
    while len(build) < len(projects):
        search(projects[i])
        # Reset if reached end without having built all
        if i == len(projects) - 1:
            i = 0
        else:
            i += 1

    return build[::-1]


def build_adjacency_list(projects, dependencies):
    a = {p: {'built': False, 'edges': []} for p in projects}
    for (k, v) in dependencies:
        a[k]['edges'].append(v)

    return a


p = ['a', 'b', 'c', 'd', 'e', 'f']
d = [('a', 'd'), ('f', 'b'), ('b', 'd'), ('f', 'a'), ('d', 'c')]

print(build_order(p, d))

